Add the initial RQL architecture

ekinanp opened the pull request:
There's still some more work that needs to be done so this should be targeted at a separate `rql` branch (didn't want to create one now when everyone's asleep). Some of this includes

* Testing the `ast.Query` type
* Implementing the walker and stree pruning (and add the HTTP endpoint)
* Finish implementing the `meta` primary
* Support metadata schemas
* Improve error-handling
* Document everything (after release)

The PR's big, but lots of it is boilerplate.

ekinanp committed:
Add the initial rql package layout and the predicate interfaces

The separate Eval* methods will let us come up with a generic predicate
expression type while also giving us compile-time type-checking where it
is needed (the compile-time type-checking ensures that something like
the Action primary doesn't accidentally get a numeric predicate). The
*InDomain methods let us properly handle negation. Both these changes result
in a simpler predicate implementation than the previous `wash find` code because
we no longer have to create custom And/Or/Not operator classes.

Note that the previous `wash find` negation code was complicated because the
predicate interface design was conflating the "check if this predicate makes
sense for this kind of entry/metadata value" [domain] and "evaluate
the predicate for this OK value" [the corresponding Eval* method]) parts
of an entry/value predicate.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the errz package and internal.NonterminalNode

NonterminalNode gives us a generic way of implementing a non-terminal
node in the AST (like e.g. Primary).

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add asttest package to centralize common test code

Most of these methods are wrappers to the *InDomain and *Eval methods.
MUM and AssertNotImplemented are a useful way to test that the generic
predicate expression class panics if one tries to call an interface
method that isn't implemented by the wrapped predicate (see the
subsequent commit).

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the generic predicate expression type

Note that all subsequent predicates and primaries will be wrapped in the
Atom type. The Atom type's a useful place to centralize the
EntryPredicate, EntrySchemaPredicate, and ValuePredicate semantics
mentioned in rql/interfaces.go. Also, the primary predicate expression
code is in expression.go.

Some of the tests use predicates that haven't yet been shown. These
tests are meant to test And/Or's truth tables for the appropriate Eval*
methods; hopefully that part is clear from the code.

Subsequent commits will include the separate predicate/primary
implementations. These implementations will have a corresponding
TestExpression_AtomAndNot test case that ensures that they can be
marshaled as atoms, that they are negated correctly and they will likely
work with the general predicate expression interface.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the action predicate

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the Boolean predicate

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the null predicate

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the numeric predicates

NumericValue is needed for the meta primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the size predicate

This acts on an entry's size attribute and, for a metadata value, on the
number of elements in the object/array

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the string predicate

StringValue's necessary for the meta primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the time predicate

TimeValue's needed for the meta primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add a base class to centralize common primary code

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the matcher package

This will slightly simplify unmarshalling all the arrays for each of the
predicates and primaries. It is a bit hacky, but it does DRY up some
redundant code.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the action primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the time attribute primaries

The test cases should be virtually identical across each primary.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the boolean primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the name and cname primaries

These tests should also be virtually identical

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the path primary

Note how we're using EntryInDomain to ensure that no path glob can ever
match the starting path (even if that path glob is negated).

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the kind primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the size primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the main ast.Query type

Tests will be added in subsequent commits. This is meant to show the RQL
AST's architecture.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on api/rql/internal/predicate/expression/expression.go:
```diff
@@ -0,0 +1,142 @@
+where we used De'Morgan's law to distribute the NOT inside the OR, and noted
+that NOT(NOT(p)) == p.
+*/
+func reduce(exp rql.ASTNode) expressionNode {
```
Note that all the custom negation and De'Morgan's law crap in `wash find` was effectively doing this to the parsed predicate expression.

ekinanp commented on api/rql/internal/predicate/action_test.go:
```diff
@@ -0,0 +1,65 @@
+	})
+
+	s.MUM(expr, "exec")
+	s.EAFTC(expr, plugin.ListAction())
```
Note that when testing `Not`, this becomes `EATTC`. This symmetry (switching `F` and `T` around) is common in all the other predicate and primary tests.

ekinanp commented on api/rql/internal/predicate/boolean_test.go:
```diff
@@ -0,0 +1,120 @@
+
+	s.MUM(expr, []interface{}{"NOT", true})
+	s.EVTTC(expr, false)
+	s.EVFTC(expr, true, "foo")
```
Notice that `NOT` still returns false for a mis-typed value b/c that value's outside the predicate's domain.

ekinanp commented on api/rql/internal/predicate/numeric.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (p *numericValue) Unmarshal(input interface{}) error {
+	if !matcher.Array(matcher.Value("number"))(input) {
```
I'm open to ideas on better centralizing all this redundant unmarshal validation. On my end, I haven't really been able to come up with a clean name for some helper/base type that does this, which is why I stuck with `matcher` as the (horrible) compromise.

MikaelSmith commented on api/rql/ast/predicateExpression.go:
```diff
@@ -0,0 +1,46 @@
+	"github.com/shopspring/decimal"
+)
+
+// PE_Primary returns a node representing PE Primary
```
PE wasn't obvious to me, maybe write out `predicate expression primary` in the comment.

MikaelSmith replied:
Just needs to be spelled out the first time it's used in a file.

ekinanp replied:
Agreed. I'll update.

ekinanp replied:
Updated.

MikaelSmith commented on api/rql/interfaces.go:
```diff
@@ -0,0 +1,93 @@
+// and EvalEntry (including negation). Note that p.EntrySchemaInDomain and
+// p.EvalEntrySchema can assume that s != nil.
+//   * If p implements EntrySchemaPredicate, then for RQL
+//         EvalEntrySchema(s) == p.EntrySchemaInDomain(s) && p.EvalEntrySchema(s)
```
I'm not clear what the evaluation order is in this expression.

ekinanp replied:
Should be `InDomain` first then the `Eval` method w/ appropriate short-circuiting. I'll clarify that.

ekinanp replied:
Updated.

MikaelSmith replied:
Parentheses would have also worked.

MikaelSmith commented on api/rql/internal/errz/matchError.go:
```diff
@@ -0,0 +1,38 @@
+
+// IsMatchError returns true if err is a MatchError,
+// false otherwise.
+func IsMatchError(err error) bool {
```
You can just use `errors.As(err, MatchError{})` in Go 1.13.

ekinanp replied:
Hmm, that doesn't quite work because it returns true if `err`'s chain contains first instance of MatchError. Some parts of the RQL code wrap match errors, see e.g. https://github.com/ekinanp/wash/blob/rql/api/rql/internal/predicate/numeric.go#L121, so that they could be returned as syntax errors. There, `errors.As` will return true.

I think `errors.As`/`errors.Is` should have been named to something like `errors.Contains`. Since `Contains` is different from `IsMatchError`, I'm opting to keep the helper.

I did find that `IsSyntaxError` is not being used anywhere, so I'll go ahead and remove that.

ekinanp replied:
Updated.

MikaelSmith commented on api/rql/internal/nonterminalNode.go:
```diff
@@ -0,0 +1,64 @@
+	SetMatchErrMsg(msg string) NonterminalNode
+}
+
+func NewNonterminalNode(n rql.ASTNode, ns ...rql.ASTNode) NonterminalNode {
```
Why two arguments? I guess that makes it so you need at least 1?

ekinanp replied:
Yup. I'll comment that.

ekinanp replied:
Updated

MikaelSmith commented on api/rql/internal/predicate/expression/atom.go:
```diff
@@ -0,0 +1,89 @@
+	"github.com/shopspring/decimal"
+)
+
+func Atom(p rql.ASTNode) rql.ASTNode {
```
I'm not clear what an Atom is in this context.

ekinanp replied:
Hmm, not sure what you mean. Are you looking for a comment like

> Atom wraps p into a type that works with predicate expressions. The RQL will use this to call p's appropriate Eval* methods, where each Eval* method implements its interface-specific semantics. For example, if p is a Primary _and_ EntryPredicate, then the returned atom's EvalEntry method will be evaluated as a.EvalEntry(e) == p.EntryInDomain(e) && p.EvalEntry(e). Front-end interfaces to the RQL should _always_ use the Atom type when testing their parsed predicates to ensure correct evaluation semantics.

Note that I plan on exporting functions like `Atom`, `Not` in the `ast` package once I get to refactoring `find`'s parsing.

ekinanp replied:
Updated since I feel like that comment would be useful anyways

MikaelSmith replied:
I'm not sure what an atom is supposed to represent.

MikaelSmith replied:
Ok, I guess I kind of understand.

MikaelSmith commented on api/rql/internal/predicate/numeric.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (p *numericValue) ValueInDomain(v interface{}) bool {
+	// TODO: Support stringified numbers?
```
I don't have a reason to do so.

MikaelSmith replied:
Or was this a reference to SI suffixes? If so, please provide an example.

ekinanp replied:
Nah, just stringified numbers. Like "1", "2", "3". Could probably wait until a user asks for it.

MikaelSmith commented on api/rql/internal/predicate/string.go:
```diff
@@ -0,0 +1,234 @@
+			StringEqual(""),
+		),
+	}
+	p.SetMatchErrMsg("must be formatted as either ['glob', <glob>], ['regex', <regex>], or ['=', <str>]")
```
Why not use `EQL` here as well? You went to the effort of making it a constant in lines 123 and 127.

ekinanp replied:
Went back to just using "=" because comparisonOp caused some annoying test failures. The constant was originally so `Time`, `Numeric` would get the right comparisonOp; that security is not needed since string equal has a helper, `StringEqual` that provides that.

ekinanp replied:
Updated

MikaelSmith commented on api/rql/internal/predicate/time.go:
```diff
@@ -0,0 +1,121 @@
+// the 'time' package
+type tm struct {
+	op ComparisonOp
+	t  time.Time
```
So this goes with absolute time rather than `time.Duration` from a reference point?

ekinanp replied:
Yes. I feel like that's more idiomatic with other filtering languages (e.g. I found "return log files that were modified after time X" to be more common).

`wash find` can still implement is past/future time semantics. We'd just create a wrapper to a TimePredicate that would add `[AND, [<=, <ref_time>], <time_predicate>]` to whatever predicate was parsed. Negation can also do a similar thing.

ekinanp replied:
If past/future time semantics become common in other front-end interfaces, RQL could probably be updated to support that so people don't have to keep writing wrappers around it.

MikaelSmith replied:
I think that makes sense.

MikaelSmith commented on api/rql/internal/primary/atime_test.go:
```diff
@@ -0,0 +1,86 @@
+	asttest.Suite
+}
+
+func (s *AtimeTestSuite) TestMarshal() {
```
Could be worth a mini test suite for atime/ctime/mtime/crtime to avoid duplicating all the same testing.

MikaelSmith replied:
You could probably make it a test suite that you instantiate with the particular primary name and type.

ekinanp replied:
Yeah, I think that's a good idea. I'll go ahead and do that.

ekinanp replied:
Updated

MikaelSmith commented:
A few tests seem to be failing.

ekinanp committed:
Fix failing string predicate and time predicate tests

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Remove errz.IsSyntaxError, make MatchError pointer

IsSyntaxError wasn't being used anywhere, and making MatchError a
pointer is more consistent with other types that implement the error
interface.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Clarify internal.NewNonterminalNode's argument structure

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add base test suite class for the time attribute primaries

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add some clarifying comments to expression.Atom

Signed-off-by: Enis Inan <enis.inan@puppet.com>

MikaelSmith commented on api/rql/internal/predicate/expression/expression.go:
```diff
@@ -0,0 +1,142 @@
+
+/*
+New returns a new predicate expression of 'ptype' predicates (PE).
+The AtomGenerator should generate "emtpy" structs representing
```
Typo in `emtpy` (should be `empty`?)

MikaelSmith commented on api/rql/internal/predicate/string.go:
```diff
@@ -0,0 +1,234 @@
+
+func (p *stringEqual) Unmarshal(input interface{}) error {
+	if !matcher.Array(matcher.Value("="))(input) {
+		return errz.MatchErrorf("must be formatted as ['%v', <str>]", "=")
```
There's no reason to do string substitution here.

MikaelSmith commented on api/rql/internal/predicate/string.go:
```diff
@@ -0,0 +1,234 @@
+	}
+	array := input.([]interface{})
+	if len(array) > 2 {
+		return fmt.Errorf("must be formatted as ['%v', <str>]", "=")
```
Same here.

MikaelSmith reviewed (approved):
Just very minor comments remaining.

ekinanp committed:
Address minor typo and unnecessary string substitution

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Updated.
